services/incentivisation/service.go

```diff
@@ -0,0 +1,518 @@
+			result[enodeID] = false
+			continue
+		}
+
+		sort.Slice(timestamps, func(i, j int) bool { return timestamps[i] < timestamps[j] })
+		timestamps = append(timestamps, uint32(now))
+		for i := 1; i < len(timestamps); i++ {
+			s.log.Info("Diff", "t1", timestamps[i], "t2", timestamps[i-1], "t1-t2", timestamps[i]-timestamps[i-1])
+
+			if timestamps[i]-timestamps[i-1] > pingIntervalAllowance {
+				result[enodeID] = false
+			}
+		}
+		if result[enodeID] {
+			s.log.Info("Node passed check", "enodeID", enodeID)
+		} else {
+			s.log.Info("Node failed check", "enodeID", enodeID)
+		}
+
+	}
+	s.log.Info("voting result", "result", result)
+	return result
+}
+
+// perform is the main loop, it posts a ping, registers with the contract, check the pings and votes
+func (s *Service) perform() error {
+	hash, err := s.postPing()
+	if err != nil {
+		s.log.Error("Could not post ping", "err", err)
+		return err
+	}
+	s.log.Info("Posted ping", "hash", hash)
+
+	err = s.FetchEnodes()
+	if err != nil {
+		return err
+	}
+
+	err = s.fetchMessages()
+	if err != nil {
+		return err
+	}
+
+	err = s.checkRegistered()
+	if err != nil {
+		s.log.Error("Could not check registered", "err", err)
+		return err
+	}
+
+	// This actually updates the session
+	newSession, err := s.newSession()
+	if err != nil {
+		s.log.Error("Could not check session", "err", err)
+		return err
+	}
+
+	if !newSession {
+		s.log.Info("Not a new session idling")
+		return nil
+	}
+
+	result := s.checkPings()
+	err = s.vote(result)
+	if err != nil {
+		s.log.Error("Could not vote", "err", err)
+		return err
+	}
+
+	// Reset whisper pings
+	s.whisperPings = make(map[string][]uint32)
+
+	return nil
+}
+
+// vote reports to the contract the decisions of the votes
+func (s *Service) vote(result map[string]bool) error {
+	var behavingNodes []gethcommon.Address
+	var misbehavingNodes []gethcommon.Address
+	auth := s.auth()
+
+	for enodeIDString, passedCheck := range result {
+		enodeID, err := hex.DecodeString(enodeIDString)
+		if err != nil {
+			return err
+		}
+		if passedCheck {
+			behavingNodes = append(behavingNodes, publicKeyBytesToAddress(enodeID))
+		} else {
+			misbehavingNodes = append(misbehavingNodes, publicKeyBytesToAddress(enodeID))
+		}
+	}
+
+	_, err := s.contract.VoteSync(&bind.TransactOpts{
+		GasLimit: gasLimit,
+		From:     auth.From,
+		Signer:   auth.Signer,
+	}, behavingNodes, misbehavingNodes)
+
+	return err
+}
+
+func (s *Service) startTicker() {
+	s.ticker = time.NewTicker(tickerInterval * time.Second)
+	s.quit = make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-s.ticker.C:
+				err := s.perform()
+				if err != nil {
+					s.log.Error("could not execute tick", "err", err)
+				}
+			case <-s.quit:
+				s.ticker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+func (s *Service) Start(server *p2p.Server) error {
+	s.log.Info("Incentivisation service started 2", "address", s.addressString(), "publickey", s.publicKeyString())
```

Comment:
What is the 2 in `Incentivisation service started 2`?

Reply:
testing docker images :) I will remove it